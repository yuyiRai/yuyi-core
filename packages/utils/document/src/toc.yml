items:
  -
    name: '@yuyi919/utils'
    href: ~/overview/sharepoint.md
    items:
      -
        name: CustomUtils
        uid: CustomUtils!
        items:
          -
            name: 'App (Namespace)'
      -
        name: EventEmitter
        uid: EventEmitter!
        items:
          -
            name: EventEmitter
            uid: 'EventEmitter!EventEmitter:class'
      -
        name: index
        uid: index!
        items:
          -
            name: 'App (Namespace)'
          -
            name: CommonDto
            uid: 'index!CommonDto:class'
          -
            name: 'DateUtils (Class)'
            uid: 'index!DateUtils:class'
          -
            name: 'DateUtils (Namespace)'
            items:
              - { name: EFormatter, uid: 'index!DateUtils.EFormatter:enum' }
          -
            name: EventEmitter
            uid: 'index!EventEmitter:class'
          -
            name: IGetPathResult
            uid: 'index!IGetPathResult:interface'
          -
            name: 'isEqual (Namespace)'
          -
            name: ITypeFilterUtils
            uid: 'index!ITypeFilterUtils:interface'
          -
            name: ITypeUtils
            uid: 'index!ITypeUtils:interface'
          -
            name: Option
            uid: 'index!Option:interface'
      -
        name: LodashExtra
        uid: LodashExtra!
        items:
          -
            name: CustomizerAdapterType
            uid: 'LodashExtra!CustomizerAdapterType:enum'
          -
            name: 'isEqual (Namespace)'
      -
        name: NodeUtils
        uid: NodeUtils!
        items:
          -
            name: Args
            uid: 'NodeUtils!Args:class'
          -
            name: ArgumentOptionsFixed
            uid: 'NodeUtils!ArgumentOptionsFixed:interface'
          -
            name: ArgumentParserFixed
            uid: 'NodeUtils!ArgumentParserFixed:interface'
      -
        name: OptionsUtils
        uid: OptionsUtils!
      -
        name: PropertyUtils
        uid: PropertyUtils!
        items:
          -
            name: CommonDto
            uid: 'PropertyUtils!CommonDto:class'
          -
            name: IGetPathResult
            uid: 'PropertyUtils!IGetPathResult:interface'
      -
        name: TestUtil
        uid: TestUtil!
      -
        name: TestUtils
        uid: TestUtils!
        items:
          -
            name: BenchmarkUtils
      -
        name: TsUtils
        uid: TsUtils!
        items:
          -
            name: Base
            uid: 'TsUtils!Base:class'
          -
            name: IFunction
            uid: 'TsUtils!IFunction:interface'
          -
            name: IKeyValueMap
            uid: 'TsUtils!IKeyValueMap:interface'
      -
        name: TypeLib
        uid: TypeLib!
        items:
          -
            name: ITypeFilterUtils
            uid: 'TypeLib!ITypeFilterUtils:interface'
          -
            name: ITypeUtils
            uid: 'TypeLib!ITypeUtils:interface'
      -
        name: WasmLoader
        uid: WasmLoader!
