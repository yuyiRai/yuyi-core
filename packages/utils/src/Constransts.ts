// tslint:disable: variable-name

declare global {
  namespace NodeJS {
    interface Global {
      __DEV__?: boolean;
    }
  }
}
var __DEV__ = process.env.NODE_ENV === 'development'
if (__DEV__) {
  global.__DEV__ = true
}

export namespace FunctionFactory {
  export type Base<Args extends any[] = any[], Result = any> = (...args: Args) => Result;

  export type ReturnType<F> = F extends Base<any, infer R> ? R : any;

  export type Bind1<A, Args extends any[] = [], Result = any> = (a: A, ...args: Args) => Result;
  export type Bind2<A, B, Args extends any[] = [], Result = any> = (a: A, b: B, ...args: Args) => Result;
  export type Bind3<A, B, C, Args extends any[] = [], Result = any> = (a: A, b: B, c: C, ...args: Args) => Result;

  export type Arg0<T> = T extends Bind1<infer T, any[]> ? T : any;
  export type Arg1<T> = T extends Bind2<any, infer T, any[]> ? T : any;
  export type Arg2<T> = T extends Bind3<any, any, infer T, any[]> ? T : any;

  export type Shift1<T, A> = T extends Bind1<A, infer Args, infer Re> ? Base<Args, Re> : Function
  export type Shift2<T, A, B> = T extends Bind2<A, B, infer Args, infer Re> ? Base<Args, Re> : Function
  export type Shift3<T, A, B, C> = T extends Bind3<A, B, C, infer Args, infer Re> ? Base<Args, Re> : Function


  // export type Bind<BindArgs extends any[], Args extends any[] = [], Result = any> = ((binds: BindArgs, ...args: Args) => Result);
  // export interface ShiftAuto<T, A = any, B = any, C = any, Re = ReturnType<T>> {
  //   <Args extends T extends Bind<[A], infer Args> ? Args : any[]>(...args: Args): Re
  //   <Args extends T extends Bind<[A, B], infer Args> ? Args : any[]>(...args: Args): Re
  //   <Args extends T extends Bind<[A, B, C], infer Args> ? Args : any[]>(...args: Args): Re
  // }
  // export type Shift<T, A extends any[]> = T extends Bind<A, infer Args, infer Re> ? Base<Args, Re> : Function
}

export namespace Constant$ {

  export const ENTRIES = Object.entries
  /**
   * * keyof PropertyDescriptor
   * writable, configurable, value, enumerable
   */
  export enum DefPropDec$$ {
    W = 'writable',
    C = 'configurable',
    V = 'value',
    E = 'enumerable',
    G = 'get',
    S = 'set'
  };
  export var KEY_PREFIX_INJECT: '__$$_' = '__$$_'
  export var KEY_STRICT: 'strict' = 'strict'
  export var KEY_CONSTRUCTOR: 'constructor' = 'constructor';
  export var KEY_PROPERTY: 'property' = 'property';
  export var KEY_PROTOTYPE: 'prototype' = 'prototype'
  export var KEY_NUM: 'number' = 'number';
  export var KEY_STR: 'string' = 'string';
  export var KEY_BOOL: 'boolean' = 'boolean';
  export var KEY_FUNC: 'function' = 'function';
  export var KEY_OBJ: 'object' = 'object';
  export var KEY_VAL: 'value' = DefPropDec$$.V;
  export var TRUE: true = true;
  export var FALSE: false = false;
  export var UNDEFINED = undefined;
  export var NULL = null;
  export var ARRAY = Array;
  export var IS_ARR = ARRAY.isArray;
  export var FUNCTION = Function;
  export var CALLER = FUNCTION.call
  export var Num = Number;
  export var REGEXP = RegExp;
  // tslint:disable-next-line: variable-name
  export var OBJECT = Object;
  export var EMPTY_OBJECT = OBJECT.seal(OBJECT.create(null))
	/**
	 * ```ts
	 * Object.keys
	 * ```
	 * */
  export var OBJ_KEYS = OBJECT.keys;
	/** 
	 * ```ts
	 * Object.freeze
	 * ``` 
	 * */
  export var OBJ_FREEZE = OBJECT.freeze;
	/**
	 * ```ts
	 * Object.assign
	 * ```
	 * */
  export var OBJ_ASSIGN = OBJECT.assign;
	/**
	 * ```ts
	 * Object.getOwnPropertyDescriptors
	 * ```
	 * */
  export var OBJ_getOwnPropertyDescriptors$ = OBJECT.getOwnPropertyDescriptors;
	/**
	 * ```ts
	 * Object.defineProperty
	 * ```
	 * */
  export var OBJ_defineProperty$ = OBJECT.defineProperty;
	/**
	 * ```ts
	 * Object.getPrototypeOf
	 * ```
	 * */
  export var OBJ_getPrototypeOf$ = OBJECT.getPrototypeOf;
	/**
	 * ```ts
	 * Object.getOwnPropertyDescriptor
	 * ```
	 * */
  export var OBJ_getOwnPropertyDescriptor$ = OBJECT.getOwnPropertyDescriptor;
	/**
	 * ```ts
	 * Object.getOwnPropertyNames
	 * ```
	 * */

  export var OBJ_getOwnPropertyNames$ = OBJECT.getOwnPropertyNames;
  export var PROMISE = Promise;
  export var CREATE_NEW = Reflect.construct
  export function CREATE_PROMISE<T>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T> {
    return CREATE_NEW(PROMISE, [executor])
  }
  export var PARSE_FLOAT = parseFloat
  export var PARSE_INT = parseInt


  export function BIND<T extends FunctionFactory.Base<[any, ...any[]], any>,
    TA,
    R extends T extends (thisArgs: any, ...other: infer K) => infer Re ? FunctionFactory.Base<K, Re> : Function
  >(Func: T, thisArgs: TA, ...args: any[]): R {
    return FUNCTION.bind.call(Func, thisArgs, ...args)
  }

  // tslint:disable-next-line: class-name
  // export type STATIC_BIND<
  //   T extends Function,
  //   Bindex extends any[]
  // > = {
    
  // }
  export function BindArg$$<
    T extends Function,
    A extends FunctionFactory.Arg0<T>
    >(Func: T, arg0: A): FunctionFactory.Shift1<T, A>
  export function BindArg$$<
    T extends Function,
    A extends FunctionFactory.Arg0<T>,
    B extends FunctionFactory.Arg1<T>,
    >(Func: T, arg0: A, arg1: B): FunctionFactory.Shift2<T, A, B>;
  export function BindArg$$<
    T extends Function,
    A extends FunctionFactory.Arg0<T>,
    B extends FunctionFactory.Arg1<T>,
    C extends FunctionFactory.Arg2<T>,
    >(Func: T, arg0: A, arg1: B, arg2: C): FunctionFactory.Shift3<T, A, B, C>;
  export function BindArg$$(Func: any, ...args: any[]): any {
    // @ts-ignore
    return BIND(Func, null, ...args)
  }

  export interface FunctionArgsShifter<
    T extends Function,
    P extends T['prototype'],
    K extends keyof P
    > {
    <Args extends P[K] extends (arg: infer A) => any ? A : any>(source: P, arg: Args): P[K] extends (...args: any[]) => infer R ? R : any;
  }

  export function INSTANCE_BIND<
    T extends Function,
    P extends T['prototype'],
    K extends keyof P,
  >(instance: T, key: K): FunctionArgsShifter<T, P, K> {
    // @ts-ignore
    return BIND(CALLER, instance[KEY_PROTOTYPE][key])
  }


  export type ForFuncCalbackFn<T, K extends TKey, R = void> = (currentValue: T, currentIndex: K, array: T[]) => R
  export type LoopCalbackFn<T, R = void> = ForFuncCalbackFn<T, number, R>

  /**
   * 原生的map循环改为函数调用
   * @param arr
   * @param callbackfn
   */
  export const MAP = INSTANCE_BIND(ARRAY, 'map') as {
    <T extends any, R = T>(arr: T[], callbackfn: LoopCalbackFn<T, R>, initialValue?: any[]): R[]
  }
  // MAP()
  /**
   * 原生的Array.prototype.some改为函数调用
   * @param arr
   * @param callbackfn
   * @param thisArg
   */
  export const SOME = INSTANCE_BIND(ARRAY, 'some') as {
    <T extends any>(arr: T[], callbackfn: LoopCalbackFn<T, boolean>, thisArg?: any): boolean
  }
  /**
   * 原生的for循环改为函数调用
   * @param arr
   * @param callbackfn
   */
  export const FOR_EACH = INSTANCE_BIND(ARRAY, 'forEach') as {
    <T extends any>(arr: T[], callbackfn: LoopCalbackFn<T, any>): void;
  }

  /**
   * 原生的for循环改为函数调用
   * @param arr
   * @param callbackfn
   */
  export const FILTER = INSTANCE_BIND(ARRAY, 'filter') as {
    <T extends any>(arr: T[], callbackfn: LoopCalbackFn<T, boolean>): T[];
  }


  /**
   * 原生的reduce改为函数调用
   * @param arr
   * @param callbackfn
   * @param initialValue
   * @example
   * REDUCE([1, 2, 3, 4], (target: any, i: number) => ({ ...target, [i]: i }), {})
   */
  export const REDUCE = INSTANCE_BIND(ARRAY, 'reduce') as {
    <T extends any, R = T>(
      arr: T[],
      callbackfn: (previousValue: R, currentValue: T, currentIndex: number, array: T[]) => R,
      initialValue?: R
    ): R
  }
  /**
   * 原生的Array<T>.push改为函数调用
   * @param arr 目标数组
   * @param elements 新元素
   * @example
   * ARR_PUSH([1, 2, 3, 4], 5, 6, 7)
   */
  export const ARR_PUSH = INSTANCE_BIND(ARRAY, 'push') as {
    <T extends any, E = T>(arr: T[], ...elements: E[]): number
  }

  /**
   * 原生的push改为函数调用
   * @param arr 目标数组
   * @param elements 新元素
   * @example
   * REGEXP_TEST(/123/, '123') // => true
   */
  export const REGEXP_TEST = INSTANCE_BIND(REGEXP, 'test') as {
    (reg: RegExp, str: string): boolean
  }

  export const delay$$ = setTimeout;

  /**
   *
   * @param arr
   * @param callbackfn
   */
  export function MAP$$<T, R>(arr: T[], callbackfn: LoopCalbackFn<T, R>): R[]
  export function MAP$$<T, R>(arr: T[], callbackfn: LoopCalbackFn<T, R>, initialValue = [], i: number = 0, length = arr.length - 1) {
    initialValue[i] = callbackfn(arr[i], i, arr)
    // @ts-ignore
    return i < length ? MAP$$(arr, callbackfn, initialValue, i + 1, length) : initialValue
  }


  /**
   *
   * @param arr
   * @param callbackfn
   */
  export function FOR_EACH$$<T>(arr: T[], callbackfn: LoopCalbackFn<T, any>): true {
    var tmp = { i: 0 }
    var length = arr.length - 1
    try {
      return FOR_EACH_P$$(arr, callbackfn, tmp, length)
    } catch (error) {
      while (length - tmp.i++) callbackfn(arr[tmp.i], tmp.i, arr)
      return true
    }
  }
  export function FOR_EACH_P$$<T>(arr: T[], callbackfn: LoopCalbackFn<T, any>, tmp: { i: number }, length: number) {
    callbackfn(arr[tmp.i], tmp.i, arr)
    // console.trace('test')
    // @ts-ignore
    return tmp.i++ < length ? FOR_EACH_P$$(arr, callbackfn, tmp, length) : true
  }
}

declare global {
  type ConstructorType<T, Args extends any[] = [any?, any?, ...any[]]> = {
    new(...args: Args): InstanceType<new (...args: any) => T>
  }
  type TKey = string | number | symbol
  type IKeyValueMap<V = any, K extends TKey = TKey> = Record<K, V>
  var K$: typeof Constant$;
}

// @ts-ignore
// tslint:disable-next-line: no-eval
// K$ = eval('Constant$')
