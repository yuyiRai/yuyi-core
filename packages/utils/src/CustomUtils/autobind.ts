import { cloneDeep } from '../LodashExtra/lodash';
import { isPureObj } from '../LodashExtra/isObject';

/**
 * 自动绑定对象中的方法的this
 * @param obj 
 * @param isCloneDeep default is true
 * @default true
 */
export function autoBindObj<T>(obj: T, isCloneDeep = true) {
  isCloneDeep && (obj = cloneDeep(obj));
  if (isPureObj(obj)) {
    Object.entries(obj).forEach(entry => {
      if (entry[1] instanceof Function) {
        obj[entry[0]] = entry[1].bind(obj);
      }
    });
  }
  return obj;
}
