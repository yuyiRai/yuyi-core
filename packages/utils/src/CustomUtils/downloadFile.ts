/* istanbul ignore next */
/**
 * 
 * @param fileId 
 * @param fileName 
 * @param getUrl 
 * @example
 * downloadFile('idxxx', 'test.xxx', id => `/api/file/get?id=${id}`)
 */
export function downloadFile(fileId: string, fileName: string, getUrl: ((id: string) => string)) {
  const a = document.createElement('a');
  a.href = getUrl(fileId);
  a.download = fileName || fileId;
  a.click();
  a.remove();
}
