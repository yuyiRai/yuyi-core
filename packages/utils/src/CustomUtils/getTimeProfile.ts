import { isNaN } from 'lodash'

export function timeProfile(): void {
  (timeProfile as any)._last = getTime()
}
export function getTimeProfile(): number {
  // tslint:disable-next-line: no-unexpected-multiline
  const a: number = (timeProfile as any)._last;
  const r = (typeof a === 'number' && !isNaN(a)) ? getTime() - a : NaN;
  (timeProfile as any)._last = NaN;
  return r
}
export function cleanTimeProfile(): void {
  // tslint:disable-next-line: no-unexpected-multiline
  (timeProfile as any)._last = NaN;
}

export function getTime() {
  return new Date().getTime()
}
