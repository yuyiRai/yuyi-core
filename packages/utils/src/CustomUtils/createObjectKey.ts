/**
 * @module CustomUtils
 */
import { values } from '../LodashExtra';
import { Constant$ } from '../Constransts';
export function createObjectKey(obj: any): string {
  const { ARR_CONCAT, OBJ_KEYS } = Constant$
  return ARR_CONCAT(OBJ_KEYS(obj), values(obj)).join();
}
