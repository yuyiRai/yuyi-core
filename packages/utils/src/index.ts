/**
 * @author yuyi
 * @packageDocumentation
 */
import './Constransts'
// import { oc } from 'ts-optchain';
// import { ArrayList } from 'typescriptcollectionsframework'
// Import here Polyfills if needed. Recommended core-js (npm i -D core-js)
// import "core-js/fn/array.find"
// ...
// import { LodashExtraUtils } from './LodashExtra'
// import '@yuyi919/env'

// const A = Type
// export const Constant$$ = { ...Constant$ }
// export const AAA = { a: { b: { c: 3 } } }
// export const AAAA = oc(AAA).a.b.c

// export const A = new ArrayList(undefined);

export * from './CustomUtils'
export * from './OptionsUtils'
export * from './TimeBuffer'
export * from './LodashExtra'
export * from './FormGirdUtils'
export * from './TypeLib'
export * from './HttpBox'
export * from './PropertyUtils'
export * from './commonUtils'
export * from './TestUtils'
export * from './ParseUtils'
export * from './MobxUtils'
export * from './EventEmitter'
export * from './WasmLoader'

// export * from './Utils'
