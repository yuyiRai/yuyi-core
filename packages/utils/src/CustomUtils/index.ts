/**
 * @module CustomUtils
 */

import { zipEmptyData } from './zipEmptyData';
export * from './pipe';
import { uuid } from './uuid';
import { downloadFile } from './downloadFile';
import { createObjectKey } from './createObjectKey';
import { getListDifferent } from './getListDifferent';
import { connectTo } from './connectTo';
// import { jsGetAge } from './jsGetAge';
import { likeArray } from './likeArray';
import { getRealLength } from './getRealLength'
import { MultipleSplit } from './MultipleSplit'
export * from './changeKeys';
export * from './waitingPromise';
export * from './autobind'

export {
  uuid,
  downloadFile,
  zipEmptyData,
  createObjectKey,
  getListDifferent,
  getRealLength,
  connectTo,
  // jsGetAge,
  likeArray,
  MultipleSplit
}
// export * from './getTimeProfile'
