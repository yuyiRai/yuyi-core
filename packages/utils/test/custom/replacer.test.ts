import { shiftKeyGroup } from '../../src'
import { shiftKeyTo, ShiftMode } from '../../src/CustomUtils/changeKeys'
// tslint:disable-next-line: no-eval
const testObj = eval(require('./test.json'))

const autotestObj = { a1: 2, a2: 4, a3: 6 }

describe('changeObjectKey', () => {
  it('shift key', () => {
    const shiftTest = { a: true, b: 1, c: '2' }
    expect(shiftKeyTo(shiftTest, 'a', 'b')).toEqual({ a: 1, b: true, c: '2' })
    expect(shiftKeyTo(shiftTest, 'a', 'b', ShiftMode.Auto)).toEqual({ a: 1, b: true, c: '2' })
    expect(shiftKeyTo(shiftTest, 'a', 'b', ShiftMode.Copy)).toEqual({ a: true, b: true, c: '2' })

    const r = shiftKeyTo.cut(shiftTest, 'a', 'b')
    expect(shiftKeyTo(shiftTest, 'a', 'b', ShiftMode.Cut)).toEqual(r)
    expect(shiftKeyTo(shiftTest, 'a', 'b', ShiftMode.Cut)).not.toBe(r)
    // @ts-ignore
    expect(r.a).toBe(undefined)
    expect(r.b).toBe(true)
    expect(r).toEqual({ b: true, c: '2' })
  })

  const testMap1 = { a1: 'a2' }
  for (const [key, title, result] of [
    [ShiftMode.Auto, 'auto模式会自动交换字段', { a1: 4, a2: 2, a3: 6 }],
    [undefined, '不设（同auto）', { a1: 4, a2: 2, a3: 6 }],
    [ShiftMode.Cut, '剪切模式，移动后删除原字段', { a2: 2, a3: 6 }],
    [ShiftMode.Copy, '复制模式，会保留原字段', { a1: 2, a2: 2, a3: 6 }]
  ]) {
    it('use replacer ' + title, () => {
      const r = shiftKeyGroup(autotestObj, testMap1, key as any)
      expect(r).not.toBe(autotestObj)
      const tmp = expect(r)
      result ? tmp.toEqual(result) : tmp.toMatchSnapshot();
    })
  }
  const testMap2 = { a1: 'a3', a2: 'a1', a3: 'a2' }
  const testMapResult = { a1: 4, a2: 6, a3: 2 }
  for (const [key, title] of [
    [ShiftMode.Auto, 'auto模式也会完全按照控制'],
    [ShiftMode.Cut, '删除模式也会完全按照控制'],
    [ShiftMode.Copy, '保留模式也会完全按照控制']
  ]) {
    it('use replacer ' + title, () => {
      expect(shiftKeyGroup(autotestObj, testMap2, key as any)).toEqual(testMapResult);
    })
  }

  it('use replacer', () => {
    const r = shiftKeyGroup(
      testObj,
      {
        '3': 7,
        '4': 5,
        '5': 6,
        '6': 3,
        '7': 4
      },
      (key, map) => {
        if (/CUST_[0-9]/.test(key)) {
          const [, num] = key.match(/CUST_([0-9])/) || []
          if (map[num]) {
            return key.replace(/CUST_([0-9])/, 'CUST_' + map[num])
          }
        }
        return null
      }
    )
    expect(r).toMatchSnapshot()
  })
})
