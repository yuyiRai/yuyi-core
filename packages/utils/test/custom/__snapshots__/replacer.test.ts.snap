// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeObjectKey use replacer 1`] = `
Object {
  "UNITAREA_1-H_NATIONALITY": Array [
    "PRK",
    "SSD",
    "CAF",
    "COD",
    "GNB",
  ],
  "UNITCUST_1-HIGH_SCORE": "4",
  "UNITCUST_1-H_UNITNATURE": Array [
    "140",
  ],
  "UNITCUST_1-LOW_SCORE": "1",
  "UNITCUST_1-L_UNITNATURE": Array [
    "110",
    "120",
    "270",
  ],
  "UNITCUST_1-MEDIUM_SCORE": "2",
  "UNITCUST_1-M_UNITNATURE": Array [
    "280",
    "300",
    "360",
    "107",
  ],
  "UNITCUST_2-HIGH_SCORE": "3",
  "UNITCUST_2-H_BUSISOURCE": Array [
    "002",
    "1403",
  ],
  "UNITCUST_2-LOW_SCORE": "1",
  "UNITCUST_2-L_BUSISOURCE": Array [
    "000",
    "100",
  ],
  "UNITCUST_2-MEDIUM_SCORE": "2",
  "UNITCUST_2-M_BUSISOURCE": Array [
    "2202",
    "2203",
  ],
  "UNITCUST_3-HIGH_SCORE": "6",
  "UNITCUST_3-LOW_SCORE": "1",
  "UNITCUST_3-MEDIUM_SCORE": "2",
  "UNITCUST_4-C_TRADECOUNT-LT-YEAR": "1",
  "UNITCUST_4-HIGH_SCORE": "5",
  "UNITCUST_4-H_TRADECOUNT-GE-TRADE_COUNT": "3",
  "UNITCUST_4-LOW_SCORE": "1",
  "UNITCUST_4-MEDIUM_SCORE": "4",
  "UNITCUST_4-M_TRADECOUNT-LT-TRADE_COUNT": "3",
  "UNITCUST_4LYEAR": "1",
  "UNITCUST_4MYEAR": "1",
  "UNITCUST_5-HIGH_SCORE": "3",
  "UNITCUST_5-H_SHAREHOLDERTYPE": Array [
    "01",
    "02",
  ],
  "UNITCUST_5-LOW_SCORE": "1",
  "UNITCUST_5-L_SHAREHOLDERTYPE": Array [
    "07",
    "08",
    "09",
    "10",
    "11",
  ],
  "UNITCUST_5-MEDIUM_SCORE": "2",
  "UNITCUST_5-M_SHAREHOLDERTYPE": Array [
    "03",
    "04",
    "05",
    "06",
  ],
  "UNITCUST_6-HIGH_SCORE": "6",
  "UNITCUST_6-H_IDCHECK": Array [
    "101",
    "104",
    "105",
    "107",
    "108",
  ],
  "UNITCUST_6-LOW_SCORE": "1",
  "UNITCUST_6-L_IDCHECK": Array [
    "100",
  ],
  "UNITCUST_6-MEDIUM_SCORE": "2",
  "UNITCUST_6-M_IDCHECK": Array [
    "102",
    "103",
    "109",
    "106",
  ],
  "UNITCUST_7-HIGH_SCORE": "3",
  "UNITCUST_7-H_COMPANY-LE-TIME": "1",
  "UNITCUST_7-LOW_SCORE": "1",
  "UNITCUST_7-L_COMPANY-GT-TIME": "3",
  "UNITCUST_7-MEDIUM_SCORE": "2",
  "UNITCUST_7-M_COMPANY-LE-TIME": "3",
  "UNITCUST_7MCOMPANYLETIME": "1",
  "UNITIDSY_1-LOW_SCORE": "1",
  "UNITIDSY_1-MEDIUM_SCORE": "3",
  "UNITIDSY_2-HIGH_SCORE": "5",
  "UNITIDSY_2-H_JOBTITLE": Array [
    "01",
    "02",
    "03",
  ],
  "UNITIDSY_2-LOW_SCORE": "1",
  "UNITIDSY_2-MEDIUM_SCORE": "2",
  "busiNo": null,
  "custType": "4",
}
`;
